### DISCUSSION TITLE
Uncaught Exception for AyncError within the AsyncNotifier `build`

### DISCUSSION THREAD

**Discussion by tadaspetra** [rrousselGit/riverpod#2523] [Q&A] [answered]
I figured this would make sense as a separate discussion. When throwing an error in response to a `build` function call, the IDE marks that as an "Uncaught Exception". Even though technically it is being handled behind the scene's, is this just something that can be avoided?

https://github.com/rrousselGit/riverpod/discussions/2517#discussioncomment-5797795

**Comment by rrousselGit** [comment]
This is voluntary (at least was), to help debug exeptions inside providers.   
That's been changed in recent versions already due to complaints though. Make sure you're up-to-date.

If you still have the problem, please make a reproducible example.

**Comment by tadaspetra** [comment]
I believe I am on the latest version `flutter_riverpod: 2.3.6` (checked the pubspec.lock to make sure), and I still am getting this as an "Uncaught Exception". 
![Screenshot 2023-05-04 at 9 37 22 AM](https://user-images.githubusercontent.com/60107328/236240874-ad4c3e12-6ae0-4751-b968-798453bd8576.png)

Here's the full build method, and function if it helps.
```dart
class UserNotifier extends AsyncNotifier<UserState?> {
  final _auth = FirebaseAuth.instance;
  final _firestore = FirebaseFirestore.instance;

  @override
  build() {
    return ref.watch(authProvider).maybeWhen(
      data: (user) {
        if (user != null) {
          return getLoggedInUser(user.email!);
        }
        return null;
      },
      orElse: () {
        return null;
      },
    );
  }

  Future<UserState> getLoggedInUser(String email) async {
    QuerySnapshot response = await _firestore
        .collection("users")
        .where('email', isEqualTo: email)
        .get();
    if (response.docs.isEmpty) {
      throw ("No firestore user associated to authenticated email");
    }
    if (response.docs.length != 1) {
      throw ("More than one firestore user associated to authenticated email");
    }
    return UserState(
      id: response.docs[0].id,
      user: FirestoreUser.fromMap(
          response.docs[0].data() as Map<String, dynamic>),
    );
  }
```

**Comment by tadaspetra** [comment]
Made a simpler example. The exception only throws when I `await` the `Future.delayed`.

```dart
class UserNotifier extends AsyncNotifier<String?> {
  @override
  build() {
    return ref.watch(authProvider).maybeWhen(
      data: (user) {
        if (user != null) {
          return getLoggedInUser("fake");
        }
        return null;
      },
      orElse: () {
        return null;
      },
    );
  }

  Future<String> getLoggedInUser(String email) async {
    await Future.delayed(const Duration(seconds: 2));
    if (email == "fake") {
      throw ("Fake email");
    }
    return "user@gmail.com";
  }
```

**Reply by tadaspetra** [reply]
Found a work around for this, that doesn't throw an exception using `then`, but would be nice to be able to use `await`

```dart
Future<String> getLoggedInUser(String email) async {
  return Future.delayed(const Duration(seconds: 2)).then((value) {
    if (email == "fake") {
      throw ("Fake email");
    }
    return "user@gmail.com";
  });
}
```

**Comment by rrousselGit** [comment]
It sounds like the issue is because the async operation is split into multiple async blocks.

In theory, throwing directly inside "build" should be fine. But for anything else, I have no control over the IDE behavior

**Comment by tadaspetra** [comment]
If awaiting inside the `build`, then throwing it still gets caught.

![Screenshot 2023-05-04 at 11 48 47 AM](https://user-images.githubusercontent.com/60107328/236271536-27e15fe4-5832-467a-b83e-e381346a7fcc.png)

**Comment by tadaspetra** [comment]
An even simpler example that gives an "Uncaught Exception"

```dart
  @override
  build() async {
    await Future.delayed(const Duration(seconds: 1));
    throw ("Error");
  }
```

**Comment by smiLLe** [comment]
I am having the same issue but i am not sure if there is a fix.

This will execute the "uncaught exception" if dio throws an error because of a bad response (400,401 .. ).
```dart
Future<Response> get(..) async {
  return await dio.get( ... );
}
```

**Reply by smiLLe** [reply, marked as answer]
@tadaspetra 
This might be working using `try catch`

```dart
Future<Response> get(..) async  {
  try {
    return await dio.get( ... );
  }
  catch (e, trace) {
   return Future<Response>.error(e, trace);
  }
}
```

**Reply by tadaspetra** [reply]
@smiLLe Yep, that seems to be working! Appreciate it 😄 

The `then()` works well too, but I prefer using async/await like your example
